### ISSUE TITLE
Computing class weights from class frequencies on training set

### ISSUE THREAD

**Issue by TimZaragori** [MIC-DKFZ/nnUNet#1902]
Dear nnUNet devs,

I would like to introduce some class weights in my loss function to see if it can help to face the imbalance between background and foreground voxels. 

I wanted your advice on the best way to retrieve this information and integrate it during loss definition in my `nnUNetTrainerV2` custom class. I am still working with nnUNet v1 as the development started before v2 release. Here are my ideas :
1. Compute it directly from the data when it is loaded in the memory before the training starts. I found where the DataLoader is created and that the data are unpacked in `nnUNetTrainerV2` to create `.npy` files but not when they are loaded in memory. Are data loaded into memory just before each iteration?
2. Add the information in the dataset_properties file so it will be included in the plans file and be easily accessible after

Thanks in advance,
Timothée

**Comment by constantinulrich** [comment]
Hi,
the dataloader loads a bunch of image label pairs and processes these images (augmentations) and holds them in a "buffer". 
During each iteration the next batch is then retrieved from the buffer. Normally, the dataloader does not load all images of the dataset at once. 

If your class weights are determined based on the whole dataset, I would recommend to do this separately on the preprocessed datafolders and then load the information you need when initializing a new trainer class. 

Best
constantin

**Comment by TimZaragori** [comment]
Ok, thanks for your quick answer!
